### ISSUE TITLE
Allow user-specified initialization function for precompilation

### ISSUE THREAD

**Issue by sbromberger** [johnmyleswhite/Benchmarks.jl#1]
Hi,

This mirrors closely what I wanted for LightGraphs. I have a suggestion, though. In https://github.com/johnmyleswhite/Benchmarks.jl/blob/master/src/Benchmarks.jl#L202, you run the benchmark functions first to pre-compile. For very expensive functions, this might not be ideal (for example, the `betweenness_centrality` benchmark I'd like to track takes ~20 minutes to run). 

Would you consider allowing the user to supply an "initialization" function that would be responsible for the precompilation? This way, I could call `betweenness_centrality` with a much smaller graph first, then use the large graph for the benchmark.

**Comment by johnmyleswhite** [comment]
Sure, I can do that. But wouldn't the rest of this code also do something crazy if the benchmark function takes 15 minutes to execute since you wouldn't be able to take many samples?

I still need to put time budgets into this code. But I'd like to understand your use case. Do you want to have a single execution of your benchmark? Or would you want the rest of this machinery?

**Comment by sbromberger** [comment]
So the issue I have is that some functions (Dijkstra, neighbors, etc.) run in micro- to millisecond time, but there are some very expensive functions like betweenness centrality that can take minutes / hours. (A 39796-node graph with 301498 edges takes about 20 mins. This graph is used to compare LightGraphs against [other packages](https://graph-tool.skewed.de/performance).) So ideally there'd be some flexibility there where I could specify "run these functions 1000 times to get a benchmark" and "run this function only once", depending on how long it takes to actually run.

In other words, a sample size of 1 is fine for very expensive functions. :)

**Comment by johnmyleswhite** [comment]
Ok. That's good to know. I need to think through that use case a little more since it will invalidate a bunch of downstream assumptions about reporting uncertainty.

**Comment by sbromberger** [comment]
No worries! I started building my own benchmarking for LightGraphs but then saw this last night/this morning and decided to suspend my efforts. One thing I do though is to pass an initialization function into the test:

```
type BMTest
    desc::AbstractString
    init::Function
    fn::Function
    args::Tuple
end
```

That way, I can do `betweenness_centrality(Graph(5,2))` to precompile before I pass the big graph in.

**Comment by johnmyleswhite** [comment]
I think there's some confusion here about what's going to be pre-compiled. The important thing given the current design is the act of precompiling a benchmarkable function, which always take in exactly two arguments of constant type: a plan and a set of samples. The body of a benchmarkable function can contain an expression like `betweenness_centrality(Graph(5,2))`, but that's not what we're trying to pre-compile. If we want to pre-compile the functions that are called inside of a benchmarkable function, we need to provide another mechanism for doing that.

I tried to do something like the `fn(args...)` approach that your type hints at, but I couldn't get it to produce ecologically valid numbers -- there was too large a performance gap between something like `sin(2.0)` and `fn = sin; fn((2.0, )...)`.

**Comment by prcastro** [comment]
What's the difference between this and `setup` used on `@benchmarkable` macro?

**Comment by johnmyleswhite** [comment]
`setup` executes arbitrary code before the main loop, so you could put code into that block to force any pre-compilation of inner functions. But it's inside the benchmarkable function, so the benchmarkable function itself can't be precompiled by `setup`.

**Comment by prcastro** [comment]
Oh, now I see.

One more question: We precompile a function so its compilation time won't affect the result of the benchmark, right? If so, why the precompilation of the benchmarkable function would change the results of the benchmark? Doesn't the benchmark occurs "inside" the benchmarkable function, so it wouldn't be affected by its compilation?

**Comment by johnmyleswhite** [comment]
That didn't seem to be true when I was working on this yesterday, but there's enough moving parts that I should check it again.

**Comment by johnmyleswhite** [comment]
Ok, I'm working on this now. @sbromberger, would you be ok if we defaulted to running your function exactly once if a one-call, one-sample execution takes longer than a fixed amount of time (say 1 minute)? We'd let you tweak that amount, but it seems like the easiest way to select a default that would behave sanely. Your estimates would be slightly inaccurate because of the included compilation time, but we have the `setup` preface to shrink the compilation time as much as possible.

**Comment by johnmyleswhite** [comment]
I implemented a first pass at a new data collection routine that tries to divide functions into three sorts:
- Those that should executed exactly.
- Those that are slow enough that we only need one call to generate a valid sample, but not so slow that we can't afford to record multiple samples.
- That that are fast enough that we should employ a sample size search procedure and call the function many times per sample.

I hope this resolves most of @sbromberger's concern, except that we have to live with some small amount of confounding between execution time and JIT compilation time.

**Comment by sbromberger** [comment]
>  @sbromberger, would you be ok if we defaulted to running your function exactly once if a one-call, one-sample execution takes longer than a fixed amount of time (say 1 minute)? 

I think that would be ok, but what happens with edge cases? Say I set this at 5m. If I have a function that runs for 4m30, how many times will it get executed?

**Comment by sbromberger** [comment]
Also, re: compilation time: I think this really affects functions that are very fast. That is, a `betweenness_centrality(Graph(5,2))` will run first time in ~210ms, then will run in ~60ms after that. The precompilation might skew the numbers there, no?

**Comment by johnmyleswhite** [comment]
In that case, it will fall into the second category. The logic there, which I'm implementing right now, is that:
- User specifies both (1) a number of samples desired (defaulting to 100) and (2) a budget of time in seconds to spend running the benchmark.
- We calculate the number of executions we could have executed within the user's time budget. If that number is smaller than the number of samples requested, we only execute as many fit within the budget. If the number is larger, we only execute the number of samples requested.

If the function is midrange-fast (210 ms is about 1,000,000x slower than the `sin` function), it will be precompiled before any samples are recorded.

Let me finish another commit, then ping you to review the exact code. That should make it easier to see if I'm missing edge cases.

**Comment by sbromberger** [comment]
This sounds perfect, and I'm happy to review the code. Why do we need the first category if we can specify both maxsamples and maxtime? That would work for long functions, too - you can just specify `maxsamples=1` for that particular benchmark, no?

**Comment by johnmyleswhite** [comment]
Made another pass. Will need to review in the morning, but should be close to right now: https://github.com/johnmyleswhite/Benchmarks.jl/commit/e852d55eb8092f472ab84cbce58f4afa81cf382c

**Comment by johnmyleswhite** [comment]
@sbromberger: If you have some time, could you review this file: https://github.com/johnmyleswhite/Benchmarks.jl/blob/ae7228dc8d5627f3d366ec7d1aaa35c6ee4e9b11/src/execute.jl

That lays out what I hope is a complete solution to your primary concerns. I wrote very extensive comments (for my own benefit), so I hope the logic should be very clear.

**Comment by sbromberger** [comment]
Much of this is confusing to me (your Julia skills dwarf mine), but a couple of comments:

1) do you need line 53 given `biased_time_ns`? That is, couldn't line 54 check against that value?
2) in fact, since you're checking the budget immediately afterwards, do you even need lines 52-56 (you'd need to change line 61 to be `remaining_time_ns = (10^9 * budget) - biased_time_ns`)

Other than that, I can't see any showstoppers....

**Comment by johnmyleswhite** [comment]
Thanks for the review.

Unfortunately there's a possibility that `biased_time_ns` and `time_used` are quite different. The obvious example where this would happen is when the core expression being evaluated is very fast, but the setup  expression is very slow. `biased_time_ns` is only supposed to include the cost of evaluating the core expression, but `time_used` will include everything including the setup and teardown. So `time_used` measures the actual time usage of the benchmarking setup, which is where I think you'd want to impose budgets.

**Comment by sbromberger** [comment]
What's included in the setup? (That is, can you give an example where time(setup) >> `biased_time_ns`?)

**Comment by johnmyleswhite** [comment]
Just to be clear (and hopefully not rude), I call that example "obvious" not because I think it should be obvious to anyone that the problem I raised could occur, but because I think there are actually other possible problems that I haven't even thought of yet -- and they'd also lead to same discrepancy.

One example of `setup` would be something like

```
setup  = :(x = rand(1_000_000))
core = :(sum(x))
```

I really want to benchmark how long `sum(x)` takes, but I need to make sure that `x` exists and that it's not global, so it has to be embedded in my "benchmarkable" function.

**Comment by sbromberger** [comment]
No offense taken at all :)

Frankly, my preference would be to just check the benchmarked code against the budget. Setup / teardown is immaterial to the overall question I'm trying to answer, which is "is the modification I'm about to commit going to adversely affect the performance of any of my core functions"? In this case, setup / teardown are of little importance to me, as they would happen regardless of the modification (ETA: and they're something I'm explicitly writing for the benchmark, so I should have a good idea about how long they're going to take).

**Comment by sbromberger** [comment]
PS: the README is awesome. Explains things very well. I'm currently running a benchmark on betweenness_centrality.

**Comment by johnmyleswhite** [comment]
Thanks, Seth. :)

**Comment by johnmyleswhite** [comment]
This is addressed by other changes that have been made
